platform/darwin/test/MGLPredicateTests.mm

```diff
@@ -0,0 +1,416 @@
+    [self testSymmetryWithFormat:@"a != nil" reverseFormat:@"nil != a" mustRoundTrip:YES];
+    
+    [self testSymmetryWithFormat:@"a < 1" reverseFormat:@"1 > a" mustRoundTrip:YES];
+    [self testSymmetryWithFormat:@"a <= 1" reverseFormat:@"1 >= a" mustRoundTrip:YES];
+    [self testSymmetryWithFormat:@"a > 1" reverseFormat:@"1 < a" mustRoundTrip:YES];
+    [self testSymmetryWithFormat:@"a >= 1" reverseFormat:@"1 <= a" mustRoundTrip:YES];
+    
+    [self testSymmetryWithFormat:@"a BETWEEN {1, 2}" reverseFormat:nil mustRoundTrip:YES];
+    [self testSymmetryWithPredicate:[NSPredicate predicateWithFormat:@"a BETWEEN %@", @[@1, @2]]
+                   reversePredicate:nil
+                      mustRoundTrip:YES];
+    XCTAssertThrowsSpecificNamed([NSPredicate predicateWithFormat:@"{1, 2} BETWEEN a"].mgl_filter, NSException, NSInvalidArgumentException);
+    NSPredicate *betweenSetPredicate = [NSPredicate predicateWithFormat:@"a BETWEEN %@", [NSSet setWithObjects:@1, @2, nil]];
+    XCTAssertThrowsSpecificNamed(betweenSetPredicate.mgl_filter, NSException, NSInvalidArgumentException);
+    XCTAssertThrowsSpecificNamed([NSPredicate predicateWithFormat:@"a BETWEEN {1}"].mgl_filter, NSException, NSInvalidArgumentException);
+    XCTAssertThrowsSpecificNamed([NSPredicate predicateWithFormat:@"a BETWEEN {1, 2, 3}"].mgl_filter, NSException, NSInvalidArgumentException);
+    
+    [self testSymmetryWithFormat:@"a IN {1, 2}" reverseFormat:@"{1, 2} CONTAINS a" mustRoundTrip:NO];
+    [self testSymmetryWithPredicate:[NSPredicate predicateWithFormat:@"a IN %@", @[@1, @2]]
+                   reversePredicate:[NSPredicate predicateWithFormat:@"%@ CONTAINS a", @[@1, @2]]
+                      mustRoundTrip:YES];
+    
+    [self testSymmetryWithFormat:@"{1, 2} CONTAINS a" reverseFormat:@"a IN {1, 2}" mustRoundTrip:NO];
+    [self testSymmetryWithPredicate:[NSPredicate predicateWithFormat:@"%@ CONTAINS a", @[@1, @2]]
+                   reversePredicate:[NSPredicate predicateWithFormat:@"a IN %@", @[@1, @2]]
+                      mustRoundTrip:NO];
+}
+
+- (void)testSymmetryWithFormat:(NSString *)forwardFormat reverseFormat:(NSString *)reverseFormat mustRoundTrip:(BOOL)mustRoundTrip {
+    NSPredicate *forwardPredicate = [NSPredicate predicateWithFormat:forwardFormat];
+    NSPredicate *reversePredicate = reverseFormat ? [NSPredicate predicateWithFormat:reverseFormat] : nil;
+    [self testSymmetryWithPredicate:forwardPredicate reversePredicate:reversePredicate mustRoundTrip:mustRoundTrip];
+}
+
+- (void)testSymmetryWithPredicate:(NSPredicate *)forwardPredicate reversePredicate:(NSPredicate *)reversePredicate mustRoundTrip:(BOOL)mustRoundTrip {
+    auto forwardFilter = forwardPredicate.mgl_filter;
+    NSPredicate *forwardPredicateAfter = [NSPredicate mgl_predicateWithFilter:forwardFilter];
+    if (mustRoundTrip) {
+        // Aggregates should round-trip, but for some reason only their format strings do.
```

Comment:
For the `BETWEEN` case, `forwardPredicateAfter` is already an aggregate expression under the hood, due to the use of braces [here](https://github.com/mapbox/mapbox-gl-native/blob/ea4b63cb511eea62bbb2454ce59837f009313aca/platform/darwin/src/NSPredicate%2BMGLAdditions.mm#L97). For the `IN` case, I’ll add `+[NSExpression expressionForAggregate:]` for consistency with `BETWEEN`.

I’m comfortable with the SDK “upgrading” `int`s to `long`s and literal collections to aggregates when round-tripping. While it’s unfortunate that Foundation makes distinctions here, it’s basically Postel’s Law: accepting either collections or aggregates in input, but only outputting aggregates; accepting either `int`s or `long`s in input, but only outputting `long`s.